# 1 상관계수 corr().unstack()
# 2 자기자신 제외 result = result[result!=]

In [31]:
from scipy.stats import shapiro 

In [44]:
from scipy.stats import shapiro , anderson

import pandas as pd 
import matplotlib.pyplot as plt
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/normal6.csv')
df

p , stats = shapiro(df['data'])
print(shapiro(df['data']))

print(anderson(df['data'].values))ㅑ

ShapiroResult(statistic=np.float64(0.9995353804150279), pvalue=np.float64(0.15046155392344363))
AndersonResult(statistic=np.float64(0.8266993530414766), critical_values=array([0.576, 0.656, 0.786, 0.917, 1.091]), significance_level=array([15. , 10. ,  5. ,  2.5,  1. ]), fit_result=  params: FitParams(loc=np.float64(299.95980319533163), scale=np.float64(5.031806887885131))
 success: True
 message: '`anderson` successfully fit the distribution to the data.')


D:\python\Lib\site-packages\scipy\stats\_axis_nan_policy.py:586: UserWarning: scipy.stats.shapiro: For N > 5000, computed p-value may not be accurate. Current N is 6000.
  res = hypotest_fun_out(*samples, **kwds)


# 단일 표본 t 검정

p > 0.05 보다 크면 귀무 165가 평균키이다  / 아니면 대립 아니다 

- 평균키는 165라고 판단할수있는가 
1. from scipy.stats import ttest_1samp (one-sample)
-> stats , p = ttest_1samp (df['height'] ,165)

In [49]:
import pandas as pd
from scipy.stats import ttest_1samp
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/height1.csv')

stats, p = ttest_1samp(df['height'], 165)
print(stats, p)

3.2017884987150644 0.0018367171548080209


# 등분산 검정

from scipy.stats import levene

In [54]:
from scipy.stats import levene

# t 검정 독립 표본 ( 표본의 2개)
- 정규성을 가지지않으면 사용할수없다
- 데이터가 정규성 가지면 모수 검정  -> 등분산 검정 => 등분산을 가지면 equal_var = True
- 데이터가 정규성을 가지않으면 비모수검정 =>  mannwhitneyu, 
- 데이터가 정규성을 가지면 - 등분산이 아닐경우  equal_var = False

- from scipy.stats import ttest_ind

In [55]:
import pandas as pd 
import matplotlib.pyplot as plt
df1 = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind1.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind2.csv')

print(df1.head())
print(df2.head())

         data
0  105.408458
1  108.722618
2   84.398634
3  100.510816
4  100.965949
         data
0   98.329474
1   96.931704
2   95.777325
3   98.744768
4  102.971042


In [56]:
from scipy.stats import shapiro
print(shapiro(df1))
print(shapiro(df2))

# 두 그룹 모두 Shapiro검정 결과 귀무가설(정규성을 가진다)을 기각 하지 못한다. 두 그룹은 정규성을 가진다.

from scipy.stats import levene
print()
print(levene(df1['data'],df2['data']))
# 두그룹은 levene 검정을 확인해 본결과 pvalue 는 0.11로 귀무가실을 기각히지 못한다. 그러므로 등분산은 가진다

from scipy.stats import ttest_ind
print()
print(ttest_ind(df1,df2,equal_var=True))

ShapiroResult(statistic=np.float64(0.9860948333328381), pvalue=np.float64(0.3796835726350978))
ShapiroResult(statistic=np.float64(0.9901827646291556), pvalue=np.float64(0.6793743269084671))

LeveneResult(statistic=np.float64(2.5337683795339547), pvalue=np.float64(0.11302904824468703))

TtestResult(statistic=array([2.76719074]), pvalue=array([0.00619015]), df=array([198.]))


In [66]:
import pandas as pd 
import matplotlib.pyplot as plt
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/scipy5.csv')

print(shapiro(df['B'].dropna()))
print(shapiro(df['A']))

# 한그룹만 정규성을 위반하도 t-test 불가능

# 맨히튼 검정   = mannwhitneyu / 월콕슨 검정 = ranksums

from scipy.stats import mannwhitneyu , ranksums
from scipy import stats
print(list(filter(lambda x : 'mann' in x , dir(stats))))
print(mannwhitneyu(df['A'],df['B'].dropna()))
print(ranksums(df['A'],df['B'].dropna()))

ShapiroResult(statistic=np.float64(0.9639432320175103), pvalue=np.float64(0.0001356810116492406))
ShapiroResult(statistic=np.float64(0.9375393672524898), pvalue=np.float64(6.175255664337946e-10))
['_mannwhitneyu', 'boltzmann', 'mannwhitneyu']
MannwhitneyuResult(statistic=np.float64(27036.0), pvalue=np.float64(0.9807458376150018))
RanksumsResult(statistic=np.float64(0.02446942170858557), pvalue=np.float64(0.9804781743503561))


In [67]:
# 두개 학급의 시험성적에 대한 데이터이다. 두 학습의 시험 평균(비모수검정의 경우 중위값)은 동일하다 말할 수 있는지 확인 하라

# t 대응  표본 ( 표본의 2개)
- 정규성을 가질떄  전 후 비교  - ttest_rel 
- 정규성을 안가질떄 -> wilcoxon

- from scipy.stats import ttest_rel
      - ttest_rel (a,b alternative ="greater")  # B공장 제품들의 rpm이 A 공장 제품의 rpm보다 크다고
- from scipy.stats import wilcoxon

In [69]:
from scipy.stats import ttest_rel

In [70]:
import pandas as pd 
import matplotlib.pyplot as plt
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/rel2.csv')

In [71]:
from scipy.stats import shapiro

before = df['before']
after = df['after']
print(shapiro(before))
print(shapiro(after))

from scipy.stats import levene
print()
print(levene(before,after))


from scipy.stats import ttest_rel
print(ttest_rel(before,after))

ShapiroResult(statistic=np.float64(0.9907895174016046), pvalue=np.float64(0.6065791087935604))
ShapiroResult(statistic=np.float64(0.9916963493176595), pvalue=np.float64(0.692386232128563))

LeveneResult(statistic=np.float64(0.06427968690211128), pvalue=np.float64(0.8000741651677916))
TtestResult(statistic=np.float64(-2.5535473487670677), pvalue=np.float64(0.011926744724546513), df=np.int64(119))


# 카이제곱 적합성 검정 
=> 관측된 분포가 특정 이론적 분포에 잘 맞는지 확인 > 주사위 확률 
from scipy.stats import chisquare

# 카이제곱 독립성 검정
=> 두 범주형 변수간의 연관성이 있는지 확인
from scipy.stats import chi2_contingency
table = pd.crosstab(df['a'],df['b'])
chi2_contingency(table)

In [72]:
from scipy.stats import chisquare

In [74]:
# 주사위 눈금의 발생확률은 1/6으로 모두 동일하다. 그러므로 각 눈금의 기댓값은 실제 발생한 모든값을 6으로 나눈 값이다.
from scipy.stats import chisquare
import pandas as pd 
import matplotlib.pyplot as plt
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/dice.csv')
df['expected'] = (df['counts'].sum()/6).astype('int')
print(chisquare(df.counts,df.expected)) 

Power_divergenceResult(statistic=np.float64(2.333333333333333), pvalue=np.float64(0.8013589222076911))


In [76]:
from scipy.stats import chi2_contingency

import pandas as pd 
import matplotlib.pyplot as plt
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/body/body.csv')
df.head()


측정나이 측정회원성별  신장 : cm  체중 : kg  체지방율 : %  이완기혈압(최저) : mmHg  \
0  59.0      M    175.0     70.6      19.2              91.0   
1  40.0      F    161.4     45.8      21.7              62.0   
2  27.0      M    176.6     84.8      19.2              87.0   
3  38.0      M    167.6     76.7      24.7              63.0   
4  21.0      M    165.2     66.2      21.5              83.0   

   수축기혈압(최고) : mmHg  악력D : kg  앉아윗몸앞으로굽히기 : cm  교차윗몸일으키기 : 회  제자리 멀리뛰기 : cm 등급  
0             150.0      40.6             12.2          30.0          179.0  C  
1             119.0      22.1             16.0          32.0          165.0  B  
2             153.0      45.9             13.2          61.0          216.0  B  
3             132.0      43.6             16.0          45.0          231.0  A  
4             106.0      33.5             10.6          46.0          198.0  C

In [77]:
# 다음 데이터는 국민 기초체력을 조사한 데이터이다. 성별과 등급이 독립적인지 검정하라

In [79]:
table = pd.crosstab(df['측정회원성별'] , df['등급'])
chi2_contingency(table)

Chi2ContingencyResult(statistic=np.float64(120.06233631119409), pvalue=np.float64(7.481892813401677e-26), dof=3, expected_freq=array([[1220.25, 1220.25, 1220.25, 1220.25],
       [2128.75, 2128.75, 2128.75, 2128.75]]))

# 이항 분포 / 베르누이 분포 

- 베르누이 분포 => 성공횟수 / 전체 횟수
- 이항 분포 n번 시도중 k번 성공할 확률

- 이항 분포 =>  from scipy.stats import binom

In [82]:
# 조사결과 70%의 성인 남성이 3년 동안에 적어도 1번 치과를 찾는다고 할때, 21명의 성인 남성이 임의로 추출되었다고 하자.
# 21명 중 16명 미만이 치과를 찾았을 확률(반올림하여 소숫점 이하 3자리)

from scipy.stats import binom

n = 21 
p = 0.7
k = 16

prob = binom.cdf(k-1,n,p)
print(prob)

0.6372880857234828


In [84]:
# 적어도 19명이 치과를 찾았을 확률(반올림하여 소숫점 이하 3자리)

n = 21 
p = 0.7
k = 19

prob = 1- binom.cdf(k-1,n,p)
print(prob)

0.027129370541188025


# 검정 통계량 계산 

In [86]:
#어느 학교에서 수학 시험을 본 학생 100명 중 60명이 60점 이상을 받았다. 
# 이 학교의 수학 시험의 평균 점수가 50점 이상인지 95%의 신뢰 수준에서 검정하려한다.

In [99]:
import numpy as np

n = 100
p_hat = 0.6
p = 0.5 
alpha  = 0.05

# 검정 통계랑 구하기  => 고정 획룰 - 정해진 확률  / 루트 ( 정해진 확률 * ( 1- 정해진확률) / 전체 )
z = round( (p_hat - p) / np.sqrt(p * (1-p) / n),5)
print(z)

from scipy.stats import norm

p = 1- norm.cdf(z)
print(p)

2.0
0.02275013194817921


# 일원 분산 on-way Anova 검정  => 
- 3개 이상의 평균이 동일한지 검정
- 정규성 , 등분산성 , 독립성 만족

- from scipy.stats import f_oneway
- 귀무 가설 = 동일하다 / 대립 : 동일하지않다

# 아노바 검정에서 정규성을 만족하지못할경우 
- 크루스칼 - 월리스 검정 
- from scipy.stats import kruskal

In [103]:
from scipy.stats import f_oneway  
# f_oneway(a,b,c)

In [106]:
from scipy.stats import kruskal
# kruskal(a,b,c)

# t 분포 

t 값구하기
- from scipy.stats import t
- 1. 분산 구하기 => std = np.std(df['height'] , ddof = 1)
  2. 신뢰수준 , 자유도 구하기 => conlevel =  0.95 / 자유도 = ddof = n -1
  3. 꼬리 값 계산 => t_value = t.ppf((1+conlevel) /2 , ddof) => (1+ 신뢰수준 / 2 , 자유도)

95% 신뢰구간 구하기
- lower = mean -  t_value * std / np.sqrt(n)
- upper = mean + t_value * std/  np.sqrt(n)

# 다중 선형 회귀 모델 
- from statsmodels.formula.api import ols
- formula = '종속변수 ~+독립변수+독립변수'
- model = ols(formula, data = df).fit()
- model.summary()
- 회귀 계수 
   - model.params['변수'])
- p값
   - model.pvalues['변수'])
- R-square
   - model.rsquared


- 귀무가설 : 선형관계가 있다 
- 대립 가설 : 선형관계가 없다 

p > 0.05 :
 - 귀무
p < 0.05 :
 - 대립

# 모델 예측 하기 
- age가 55, Cholesterol가 72.6일때 위 모델을 기반으로 weight값을 예측하라.
 - pred = pd.DataFrame({'age' : [55] , 'Cholesterol' : [72.6]})
 - model.predict(pred)

In [127]:
from statsmodels.formula.api import ols
import pandas as pd 
df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/28/p7.csv')
df.head()

# age와 Cholesterol을 가지고 weight를 예측하는 선형 회귀 모델을 만들려고한다. age의 회귀 계수를 구하여라

formula = 'weight ~+age+Cholesterol'
model = ols(formula, data = df).fit()
print(model.summary())
print(model.params['age'])
print(model.pvalues['age'])
print(model.rsquared)


                            OLS Regression Results                            
Dep. Variable:                 weight   R-squared:                       0.044
Model:                            OLS   Adj. R-squared:                  0.038
Method:                 Least Squares   F-statistic:                     7.574
Date:                Sun, 01 Jun 2025   Prob (F-statistic):           0.000608
Time:                        15:59:15   Log-Likelihood:                -1374.9
No. Observations:                 333   AIC:                             2756.
Df Residuals:                     330   BIC:                             2767.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      74.8953      4.455     16.813      

In [136]:
import pandas as pd
pred = pd.DataFrame({'age' : [55] ,'Cholesterol' : [72.6]}) 
model.predict(pred)

0    78.85771
dtype: float64

# 로지스틱  회귀 모델 
- from statsmodels.formula.api import logit
- formula = '종속변수 ~+독립변수+독립변수'
- model = logit(formula, data = df).fit()
- model.summary()
- 회귀 계수 
   - model.params['변수'])
- p값
   - model.pvalues['변수'])
- 오즈비 구하기
   - np.exp(model.params['변수'])
- 잔차 이탈도 구하기 (residual deviance)
   -  -2 * model.llf
- 로그 우도 구하기
  - model.llf
 
- 오류율 구하기
   - from sklearn.metrics import accuracy_score
   - error_rate = 1 - accuracy_score(test['target'] ,pred)

-  유의미하지 않은 변수
-  if p < 0.05
   - 유의미 하다
- else
   - 유의미 하지않다
 
# 예측 모델 만들기 

- 위에서 계산한 모델을 이용하여 뇌크기 90 키 70 몸무게 150일 때 지능을 예측하라
- result = pd.DataFrame({'Brain'[90] , 'Height' : [70], 'Weight' : [150]})
- result = predict(result)

In [138]:
import pandas as pd 
df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e7_p3_t.csv')
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca  thal  target  
0    3.0  0.0   6.0       0  
1    2.0  3.0   3.0       1  
2    2.0  2.0   7.0       1  
3    3.0  0.0   3.0       0  
4    1.0  0.0   3.0       0

In [139]:
# train 데이터는 앞의 210개 행을, test데이터는 나머지 부분을 사용한다

In [151]:
import pandas as pd
import numpy as np
from statsmodels.formula.api import logit 

train = df.iloc[:210].reset_index(drop=True)
test = df.iloc[210:].reset_index(drop=True)

formula = ('target ~ age+sex+cp+trestbps+chol+fbs+restecg+thalach+exang+oldpeak+slope+ca+thal')
model = logit(formula, data= train).fit()
print(model.summary())

Optimization terminated successfully.
         Current function value: 0.320236
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                 target   No. Observations:                  210
Model:                          Logit   Df Residuals:                      196
Method:                           MLE   Df Model:                           13
Date:                Sun, 01 Jun 2025   Pseudo R-squ.:                  0.5350
Time:                        16:11:30   Log-Likelihood:                -67.250
converged:                       True   LL-Null:                       -144.61
Covariance Type:            nonrobust   LLR p-value:                 2.308e-26
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -10.7525      3.655     -2.942      0.003     -17.917      -3.588
age           -0.0056      0.

# F검정

- 정상 / 환자에 대한 로그 리지스틴에 분산의 차이가 있는지 구하기
  import numpy as np

  df['log_resistin'] = np.log(df['Resistin'])

df1 = df[df['Classification'] == 1]['log_resistin']
df2 = df[df['Classification'] == 2]['log_resistin']


dfv1 = df1.var()
dfv2  =df2.var()

df1_dof = dfv1.var()-1
df2_dof = dfv2.var()-1


print(df1_dof,df2_dof)

f_stat = df2_dof / df1_dof


# 합동 분산 추정량

n1 = len(result1)
n2 = len(result2)

total_var = ((n1-1) * dfv1 + (n2-1) * dfv2) / (n1+n2 -2)


# 리지스틴 값에 유의미한 차이가 있는지 독립 표본 t 검정 

from scipy.stats import ttest_ind


result = ttest_ind(df1,df2, equal_var = True)

















  